File: RemoveRecursionRefactoring/src/ro/pub/cs/diploma/passes/ReplaceForEachLoopsWithIteratorForLoops.java
package ro.pub.cs.diploma.passes;

import com.intellij.psi.PsiForeachStatement;
import com.intellij.psi.PsiMethod;
import org.jetbrains.annotations.NotNull;
import ro.pub.cs.diploma.Refactorings;
import ro.pub.cs.diploma.Util;

import java.util.List;

public class ReplaceForEachLoopsWithIteratorForLoops implements Pass<PsiMethod, List<PsiForeachStatement>, Object> {
  @NotNull private final PsiMethod myMethod;

  private ReplaceForEachLoopsWithIteratorForLoops(@NotNull PsiMethod method) {
    myMethod = method;
  }

  @NotNull
  public static ReplaceForEachLoopsWithIteratorForLoops getInstance(@NotNull PsiMethod method) {
    return new ReplaceForEachLoopsWithIteratorForLoops(method);
  }

  @Override
  public List<PsiForeachStatement> collect(PsiMethod method) {
    return Util.INSTANCE.getPsiForEachStatements(method);
  }

  @Override
  public Object transform(List<PsiForeachStatement> statements) {
    statements.forEach(statement -> Refactorings.replaceForEachLoopWithIteratorForLoop(statement, myMethod));
    return null;
  }
}

File: RemoveRecursionRefactoring/src/ro/pub/cs/diploma/passes/ReplaceForEachLoopsWithIteratorForLoops.kt
package ro.pub.cs.diploma.passes

import com.intellij.psi.PsiForeachStatement
import com.intellij.psi.PsiMethod
import ro.pub.cs.diploma.Refactorings
import ro.pub.cs.diploma.getPsiForEachStatements

class ReplaceForEachLoopsWithIteratorForLoops(private val myMethod: PsiMethod) : Pass<PsiMethod, List<PsiForeachStatement>, Nothing?> {

  override fun collect(method: PsiMethod): List<PsiForeachStatement> = method.getPsiForEachStatements()

  override fun transform(statements: List<PsiForeachStatement>): Nothing? {
    statements.forEach { statement -> Refactorings.replaceForEachLoopWithIteratorForLoop(statement, myMethod) }
    return null
  }
}
